gwpy/io/ligolw.py

```diff
@@ -23,25 +23,89 @@
 """
 
 import os.path
+import re
 from contextlib import contextmanager
 from functools import wraps
+from importlib import import_module
+
+# note: because the "future" module provides a builtins namespace,
+# we need to do this the other way around compared to normal
+try:
+    import __builtin__ as builtins
+except ImportError:  # python >= 3
+    import builtins
 
 from six import string_types
 
 import numpy
 
+try:
+    from ligo.lw.ligolw import (
+        ElementError as LigolwElementError,
+        LIGOLWContentHandler,
+    )
+except ImportError:  # no ligo.lw
+    LigolwElementError = None
+    LIGOLWContentHandler = None
+
 from .utils import (file_list, FILE_LIKE)
 
 __author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'
 
+# record actual import function
+_real_import = builtins.__import__
+
+# XML elements
 XML_SIGNATURE = b'<?xml'
 LIGOLW_SIGNATURE = b'<!doctype ligo_lw'
 LIGOLW_ELEMENT = b'<ligo_lw>'
 
+LIGO_LW_MODULE_REGEX = re.compile(r"\Aligo.lw")
+
+# regex to catch all errors associated with reading an old-format file
+# with the new library
+_compat_errors = {
+    r"invalid Column \'process_id\'",
+    r"invalid type \'ilwd:char\'"
+}
+LIGO_LW_COMPAT_ERROR = re.compile("({})".format("|".join(_compat_errors)))
+
+
+# -- import hackery to support ligo.lw and glue.ligolw concurrently -----------
+
+def _ligolw_compat_import(name, *args, **kwargs):
+    name = LIGO_LW_MODULE_REGEX.sub("glue.ligolw", name)
+    return _real_import(name, *args, **kwargs)
+
+
+def _is_glue_ligolw_object(obj):
+    if not isinstance(obj, type):
+        obj = type(obj)
+    return str(obj.__module__).startswith("glue.ligolw")
+
+
+def ilwdchar_compat(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        use_compat = kwargs.pop(
+            "ilwdchar_compat",
+            any(map(
+                _is_glue_ligolw_object,
+                list(args) + list(kwargs.values()),
+            )),
+        )
+        try:
+            if use_compat:
+                builtins.__import__ = _ligolw_compat_import
```

Comment:
@duncanmmacleod, I don't have any real notes here, except to say that I've never seen anything like this before and I'm learning new things about what python does under-the-hood by reviewing your code, which is a lot of fun

Reply:
It's quite a neat solution to the problem of supporting two different upstream APIs with a consistent single downstream API.